src/Orleans.Runtime/GrainDirectory/GrainLocator.cs

```diff
@@ -76,10 +76,13 @@ public bool TryLocalLookup(GrainId grainId, out List<ActivationAddress> addresse
             return false;
         }
 
-        public Task UnregisterMany(List<ActivationAddress> addresses, UnregistrationCause cause)
+        public async Task UnregisterMany(List<ActivationAddress> addresses, UnregistrationCause cause)
         {
             var grainAddresses = addresses.Select(addr => ConvertToGrainAddress(addr)).ToList();
-            return this.grainDirectory.UnregisterMany(grainAddresses);
+            await this.grainDirectory.UnregisterMany(grainAddresses);
+
+            foreach (var address in addresses)
+                this.cache.Remove(address.Grain);
```

Comment:
If UnregisterMany throws, the cache wont be updated - but some or all of the registrations may have been unregistered.

We could choose to ignore this because cache will be cleaned up eventually by other means, or we could put the foreach into a finally block

Reply:
Thanks. Did that and also updates the single version of `Unregister` the same way.